### PR TITLE
Add `map_word` and `parabolic_subgroup` for Weyl groups

### PR DESCRIPTION
**Pull request by TWiedemann**
Adds `parabolic_subgroup` as discussed in #4483. This relies on the new `map_word`. I also added `parabolic_subgroup_with_projection` (maybe the name is not good) which, in addition, returns the projection homomorphism from the large Weyl group to the parabolic subgroup.

I set this to draft because:
1. Someone will have to tell me where precisely to put the exports. I assume they go to experimental/LieAlgebras/src/exports.jl. For the moment I put them there at the end of file (where I will sort them alphabetically once this is clarified).
2. Examples/jldoctests for parabolic_subgroup(_with_projection) are still missing.

### CODE REVIEW

**Comment by TWiedemann** [comment]
In my function `map_word(::WeylGroupElem, ...)`, I had to call `map_word(::Vector{Int}, ...)` with `Oscar.map_word` while `map_word(::Vector{Int}, ...)` worked fine on the REPL. Not sure what is going on there. Probably related to the doctest failure:

```
WARNING: using LieAlgebras.map_word in module Oscar conflicts with an existing identifier.
[...]
ERROR: MethodError: no method matching map_word(::WeylGroupElem, ::Vector{Int64})
```

**Comment by thofma** [comment]
You need to add `map_word` here: https://github.com/TWiedemann/Oscar.jl/blob/tw/weyl_parabolics/experimental/LieAlgebras/src/LieAlgebras.jl#L72

File: experimental/LieAlgebras/src/exports.jl [hunk 1]
```diff
@@ -51,3 +51,7 @@ export tensor_power
 export tensor_product_decomposition
 export trivial_module
 export universal_enveloping_algebra
+
+export parabolic_subgroup
+export parabolic_subgroup_with_projection
+export map_word
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/exports.jl]: This file is the right place for exports. But  you don't need to `export map_word` here, instead you need to make sure to `import` it from outside (@thofma already pointed out the place were to do it). Then the `map_word` method you define will be a method for `Oscar.map_word`, instead of defining a new unrelated function `Oscar.LieAlgebras.map_word` (more precisely, if you `import map_word` from Oscar then `Oscar.map_word` will be identical to `Oscar.LieAlgebras.map_word`)

File: experimental/LieAlgebras/test/WeylGroup-test.jl [hunk 2]
```diff
@@ -135,4 +135,97 @@
...
+        end
+        # Test that proj is the left-inverse of emb
+        for _ in 1:5
+          p = rand(para)
+          @test proj(emb(p)) == p
+        end
+      end
+    end
+  end
+end
```
Comment (fingolfin) [inline experimental/LieAlgebras/test/WeylGroup-test.jl]: Please add a trailing line

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 3]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
 
   return MapFromFunc(W, G, iso, isoinv)
 end
+
+@doc raw"""
+    map_word(w::WeylGroupElem, genimgs::Vector; init = nothing)
+
+If `init` is `nothing` and `word(w) = [UInt(`$i_1$`), ..., UInt(`$i_n$`)]`, then return the product $R_1 R_2 \cdots R_n$
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: I would omit the `UInt` for clarity
```suggestion
If `init` is `nothing` and `word(w) = [`$i_1$`, ..., `$i_n$`]`, then return the product $R_1 R_2 \cdots R_n$
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 4]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+The length of `genimgs` must be equal to the rank of the parent of `w`.
+In particular, it must not be empty.
+If `w` is the trivial element, then `init` is returned if it is different
+from `nothing`, and otherwise `one(genimgs[1])` is returned.
+
+See also: [`map_word(::Union{FPGroupElem, SubFPGroupElem}, ::Vector)`](@ref),
+[`map_word(::Union{PcGroupElem, SubPcGroupElem}, ::Vector)`](@ref).
+Note that `map_word(::WeylGroupElem)` does not provide a `genimgs_inv` parameter
+because the generators of a Weyl group are always self-inverse.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: I think this is a mistake, as it precludes writing generic code.

Instead I would accept it anyway, but you can e.g. ignore it:
```suggestion
Note that `map_word(::WeylGroupElem)` accepts but ignores the `genimgs_inv`
keyword argument because the generators of a Weyl group are always self-inverse.
```
(One could also check if `genimgs == genimgs_inv` if both are given, but I wouldn't bother)

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 5]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+"""
+function map_word(w::WeylGroupElem, genimgs::Vector; init=nothing)
+  @req length(genimgs) == number_of_generators(parent(w)) "Length of vector of images does not equal rank of Weyl group"
+  return Oscar.map_word([Int(i) for i in word(w)], genimgs, init=init)
+end
+
+@doc raw"""
+    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem) -> WeylGroup, Map{WeylGroup, WeylGroup}
+
+Returns two objects: The subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`, whose generators are set to be this vector and which is itself a `WeylGroup`, and the embedding of `U` into `W`. The elements of `vec` must be pairwise distinct integers in `1:number_of_generators(W)` and `vec` must be non-empty.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: Please wrap lines at something between 80 and 100 columns wide (it helps with reviewing changes made to this text in the future, as the diffs are nicer)

Comment (fingolfin) [reply experimental/LieAlgebras/src/WeylGroup.jl]: - it should be "Return" not "Returns" (see point 2 at <https://docs.julialang.org/en/v1/manual/documentation/>)
- I'd say the return group is *not* that subgroup `U`, rather it is a free standing Weyl group isomorphic to `U`, and return map is an isomorphism between it and `U`.

How about this:
```suggestion
Return a Weyl group `P` and an embedding $f:P\to W$ such that $f(P)$ is
the subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`.
The elements of `vec` must be pairwise distinct integers in
`1:number_of_generators(W)` and `vec` must be non-empty.
```



Oh and an example in form of `jldoctest` would be super nice :-)

Comment (fingolfin) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Probably should also say that `f` maps the i-th generator of P t the i-th generator of `U` or so (at least I assume that's what it does?)

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 3 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: I thought about that as well, but the trade-off between clarity and correctness wasn't evident for me. Anyway, I'll remove it.

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 6]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+@doc raw"""
+    map_word(w::WeylGroupElem, genimgs::Vector; init = nothing)
+
+If `init` is `nothing` and `word(w) = [UInt(`$i_1$`), ..., UInt(`$i_n$`)]`, then return the product $R_1 R_2 \cdots R_n$
+with $R_j =$ `genimgs[`$i_j$`]`.
+Otherwise return the product $xR_1 R_2 \cdots R_n$ where $x =$ `init`.
+
+The length of `genimgs` must be equal to the rank of the parent of `w`.
+In particular, it must not be empty.
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]: These two sentences imply that `map_word` would not work for a rank 0 Weyl group. 
Could you please make sure that it still works in this case by just returning `init` (in case that is set). This would make it analogous to `map_word(one(free_group), []; ...)`

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 7]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+```jldoctest
+julia> W = weyl_group(:B, 3); imgs = [2, 3, 5];
+
+julia> map_word(one(W), imgs)
+1
+
+julia> map_word(W([1]), imgs)
+2
+
+julia> map_word(W([1]), imgs, init=7)
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
julia> map_word(W([1]), imgs; init=7)
```
for enhanced readability

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 8]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+julia> map_word(W([2,1,2]), imgs) # W([2,1,2]) == W([1,2,1])
+12
+
+julia> map_word(W([3, 2, 1, 3, 2, 3]), imgs)
+2250
+```
+"""
+function map_word(w::WeylGroupElem, genimgs::Vector; init=nothing)
+  @req length(genimgs) == number_of_generators(parent(w)) "Length of vector of images does not equal rank of Weyl group"
+  return Oscar.map_word([Int(i) for i in word(w)], genimgs, init=init)
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
  return map_word(Int.(word(w)), genimgs; init=init)
```
this should help a bit with type stability for the empty word

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 9]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+end
+
+@doc raw"""
+    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem) -> WeylGroup, Map{WeylGroup, WeylGroup}
+
+Returns two objects: The subgroup `U` of `W` generated by `[inv(w)*u*w for u in gens(W)[vec]]`, whose generators are set to be this vector and which is itself a `WeylGroup`, and the embedding of `U` into `W`. The elements of `vec` must be pairwise distinct integers in `1:number_of_generators(W)` and `vec` must be non-empty.
+"""
+function parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem=one(W))
+  @req length(unique(vec)) == length(vec) "Elements of vector are not pairwise distinct"
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
  @req allunique(vec) "Elements of vector are not pairwise distinct"
  @req all(i -> 1 <= i <= number_of_generators(W), vec) "Invalid indices"
```
better to check that as well, should be cheap

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 10]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+2250
+```
+"""
+function map_word(w::WeylGroupElem, genimgs::Vector; init=nothing)
+  @req length(genimgs) == number_of_generators(parent(w)) "Length of vector of images does not equal rank of Weyl group"
+  return Oscar.map_word([Int(i) for i in word(w)], genimgs, init=init)
+end
+
+@doc raw"""
+    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem) -> WeylGroup, Map{WeylGroup, WeylGroup}
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
    parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem=one(W)) -> WeylGroup, Map{WeylGroup, WeylGroup}
```
add the default value here as well

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 11]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+"""
+function parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true)
+  if check_factor
+    # Check that every generator in gens(W)[vec] commutes with every other generator.
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req coxeter_matrix_entry_from_cartan_matrix(cm, i, j) == 2 "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
+      end;
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
      end
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 12]
```diff
@@ -257,3 +257,97 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+If `U` is not a factor of `W`, an error occurs.
+"""
+function parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true)
+  if check_factor
+    # Check that every generator in gens(W)[vec] commutes with every other generator.
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req coxeter_matrix_entry_from_cartan_matrix(cm, i, j) == 2 "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
        @req is_zero_entry(cm, i, j) "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
Is this mathematically the same? (I am not sure atm.) If yes, this would be a bit more performant

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 6 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Are rank 0 weyl groups supported? I cannot find a way to create one, which is why I wrote it this way. (Anyway, no harm in returning init in this case anyway.)

File: experimental/LieAlgebras/src/exports.jl (hunk 1 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/exports.jl]: Done

File: experimental/LieAlgebras/test/WeylGroup-test.jl (hunk 2 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/test/WeylGroup-test.jl]: Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 6 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/WeylGroup.jl]:
```
 weyl_group(zero_matrix(ZZ, 0, 0))
```
I think this is not advertised anywhere, but useful in tests to catch some edge-cases

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 4 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Makes sense! Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 7 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 8 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 10 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 9 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Done

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 12 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: I'm pretty sure it's mathematically the same (`cm(i, j) == 0` iff the simple roots with indices `i` and `j` are orthogonal iff $s_i$ commutes with $s_j$), so I made this change. Also wrapped the line here.

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 6 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: I see. I'll change the documentation. The code can stay the same because `map_word` for integer vectors has the desired behaviour.

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 5 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Done. The statement that $f(P)$ is generated by `[inv(w)*u*w for u in gens(W)[vec]]` is now technically redundant because it follows from the description of `f` on the generators, but I left it in because it seems more clear to me that way.

**Comment by TWiedemann** [comment]
@thofma @fingolfin @lgoettgens Thanks for the help and suggestions! I fixed everything.

One non-nightly test fails: Am I not allowed to spread a string over several lines with "\", as in WeylGroup.jl:398?
```
        @req is_zero_entry(cm, i, j) begin
          "Input vector does not describe a factor of the Weyl group, \
          so there is no projection homomorphism"
        end
```
[This was apparently introduced in Julia 1.7](https://discourse.julialang.org/t/single-string-on-multiple-lines-without-n/18099/7), which might be the problem.

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 13]
```diff
@@ -257,3 +257,159 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+)
+  if check_factor
+    # Check that every generator in gens(W)[vec] commutes with every other generator.
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req is_zero_entry(cm, i, j) begin
+          "Input vector does not describe a factor of the Weyl group, \
+          so there is no projection homomorphism"
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
          "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
does it work like this? I know that the line is then technically overly long, but that would be fine here IMO

**Comment by lgoettgens** [comment]
Could you please also use `map_word` in the `isomorphism` functions in the same file (cf. https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1922579409)? This would then provide as another indirect test of the newly added function

**Comment by TWiedemann** [comment]
> Could you please also use `map_word` in the `isomorphism` functions in the same file (cf. [#4478 (comment)](https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1922579409))? This would then provide as another indirect test of the newly added function

Done. Also added links to the new `map_word` in the docs of the old `map_word` functions.

I have removed the draft status.

**Review by lgoettgens — APPROVED** [review]
Looks good, thanks! @fingolfin could you please have a look again as well?

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 14]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+  W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true
+)
+  if check_factor
+    # Check that every generator in gens(W)[vec] commutes with every other generator.
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
+        @req is_zero_entry(cm, i, j) begin
+          "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
          "Input vector must describe a direct factor of the Weyl group"
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 15]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+"""
+function parabolic_subgroup_with_projection(
+  W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true
+)
+  if check_factor
+    # Check that every generator in gens(W)[vec] commutes with every other generator.
+    # In other words, vec describes a union of irreducible components of the Coxeter diagram.
+    cm = cartan_matrix(root_system(W))
+    for i in vec
+      for j in setdiff(1:number_of_generators(W), vec)
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: This recomputes `setdiff` on each iteration, can be avoided by switching the loops:
```suggestion
    for i in setdiff(1:number_of_generators(W), vec)
      for j in vec
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 16]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+julia> P2, f2, p2 = parabolic_subgroup_with_projection(W, [4,6,5])
+(Weyl group of root system of type B3 (non-canonical ordering), Map: P2 -> W, Map: W -> P2)
+
+julia> p2(W[5]) == P2[3]
+true
+```
+"""
+function parabolic_subgroup_with_projection(
+  W::WeylGroup, vec::Vector{<:Integer}; check_factor::Bool=true
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: We often have optional checks and they almost always are controlled by a kwarg `check`. Perhaps stick with that here, too?
```suggestion
  W::WeylGroup, vec::Vector{<:Integer}; check::Bool=true
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 17]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+@doc raw"""
+    parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}) -> WeylGroup, Map{WeylGroup, WeylGroup}, Map{WeylGroup, WeylGroup}
+
+Return a triple `(P, emb, proj)` that describes a factor of `W`, that is,
+a product of irreducible factors.
+Here `P, emb = `[`parabolic_subgroup`](@ref)`(W, vec)`
+and `proj` is the projection map from `W` onto `P`,
+which is a left-inverse of `emb`.
+If `P` is not a factor of `W`, an error occurs.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]: Maybe document the optional kwarg `check_factor` / `check`? Or is it deliberately undocumented?

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 18]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+julia> f3(P3[2]) == W[3]
+true
+```
+"""
+function parabolic_subgroup(W::WeylGroup, vec::Vector{<:Integer}, w::WeylGroupElem=one(W))
+  @req allunique(vec) "Elements of vector are not pairwise distinct"
+  @req all(i -> 1 <= i <= number_of_generators(W), vec) "Invalid indices"
+  cm = cartan_matrix(root_system(W))[vec, vec]
+  para = weyl_group(cm)
+  genimgs = map(x -> inv(w) * x * w, gens(W)[vec])
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
  genimgs = [conj(W[i], w) for i in vec]
```

File: experimental/LieAlgebras/test/WeylGroup-test.jl [hunk 19]
```diff
@@ -135,4 +135,99 @@
...
+          para, emb, proj = parabolic_subgroup_with_projection(W, vec)
+        else
+          para, emb = parabolic_subgroup(W, vec)
+        end
+        genimgs = [gen(W, i) for i in vec] # Desired images of gens(para) in W
+      else
+        # On subsequent runs, conjugate by random elements
+        conj = rand(W)
+        para, emb = parabolic_subgroup(W, vec, conj)
+        genimgs = [inv(conj) * gen(W, vec[i]) * conj for i in 1:length(vec)]
```
Comment (fingolfin) [inline experimental/LieAlgebras/test/WeylGroup-test.jl]:
```suggestion
        r = rand(W)
        para, emb = parabolic_subgroup(W, vec, r)
        genimgs = [conj(W[i], r) for i in vec]
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 20]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+          "Input vector does not describe a factor of the Weyl group, so there is no projection homomorphism"
+        end
+      end
+    end
+  end
+
+  factor, emb = parabolic_subgroup(W, vec)
+  # Generators of W are mapped to the corresponding generators of factor,
+  # or to 1 if there is no corresponding generator
+  proj_gen_imgs = [one(factor) for _ in 1:number_of_generators(W)]
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
  proj_gen_imgs = fill(one(factor), number_of_generators(W))
```
or (my personal preference but it's a matter of taste)
```suggestion
  proj_gen_imgs = fill(one(factor), ngens(W))
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 21]
```diff
@@ -257,3 +257,158 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+The length of `genimgs` must be equal to the rank of the parent of `w`.
+If `w` is the trivial element, then `init` is returned if it is different
+from `nothing`, and otherwise `one(genimgs[1])` is returned if `genimgs` is non-empty.
+If `w` is trivial, `init` is nothing and `genimgs` is empty, an error occurs.
+
+See also: [`map_word(::Union{FPGroupElem, SubFPGroupElem}, ::Vector)`](@ref),
+[`map_word(::Union{PcGroupElem, SubPcGroupElem}, ::Vector)`](@ref).
+Note that `map_word(::WeylGroupElem)` accepts but ignores the `genimgs_inv` keyword argument
+because the generators of a Weyl group are always self-inverse.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
Note that `map_word(::WeylGroupElem)` accepts the `genimgs_inv` keyword argument
for consistency with other `map_word` methods, but ignores it because the
generators of a Weyl group are always self-inverse.
```

**Review by fingolfin — COMMENTED** [review]
Looks good to me in principle, but I have a few small suggestions.

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 20 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Note that there is a semantical difference here. `fill` uses the same object for all positions, while the list comprehension uses different (but identical) objects.

So the former is slightly more performant, but may not be used in contexts with mutating functions. But as far as I can see that is not the case here

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 17 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: It was a deliberate choice to not document because it seemed to me that the user should not normally disable the check. I had the `set_properties` argument of `isomorphism` in mind ([previous disucssion](https://github.com/oscar-system/Oscar.jl/pull/4478#discussion_r1920039870)); however, I do see now that the situations are rather different: `set_properties = false` really only makes sense for test purposes, which is not the case for `check` here. So yes, I agree that this should be documented. Will do that.

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 16 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Makes sense

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 20 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Ah, good. I had wondered whether there was a better way to initialise the vector, thanks.

**Comment by TWiedemann** [comment]
@fingolfin Thanks! I will implement the remaining suggestions tomorrow.

**Comment by TWiedemann** [comment]
Fixed the last two issues (`check_factor` -> `check` and added documentation for `check`).

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 22]
```diff
@@ -257,3 +257,160 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+  para = weyl_group(cm)
+  genimgs = [conj(W[i], w) for i in vec]
+  emb = function (u::WeylGroupElem)
+    return map_word(u, genimgs)
+  end
+  return para, MapFromFunc(para, W, emb)
+end
+
+@doc raw"""
+    parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}) -> WeylGroup, Map{WeylGroup, WeylGroup}, Map{WeylGroup, WeylGroup}
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl]:
```suggestion
    parabolic_subgroup_with_projection(W::WeylGroup, vec::Vector{<:Integer}; check::Bool=true) -> WeylGroup, Map{WeylGroup, WeylGroup}, Map{WeylGroup, WeylGroup}
```

File: experimental/LieAlgebras/src/WeylGroup.jl [hunk 23]
```diff
@@ -257,3 +257,160 @@ function isomorphism(::Type{PermGroup}, W::WeylGroup; set_properties::Bool=true)
...
+
+Return a triple `(P, emb, proj)` that describes a factor of `W`, that is,
+a product of irreducible factors.
+Here `P, emb = `[`parabolic_subgroup`](@ref)`(W, vec)`
+and `proj` is the projection map from `W` onto `P`,
+which is a left-inverse of `emb`.
+If `check = true`, then it is checked whether `vec` actually describes
```
Comment (lgoettgens) [inline experimental/LieAlgebras/src/WeylGroup.jl:367]:
```suggestion

If `check = true`, then it is checked whether `vec` actually describes
```
new paragraph for better readability.

File: experimental/LieAlgebras/src/WeylGroup.jl (hunk 22 above)
Comment (TWiedemann) [reply experimental/LieAlgebras/src/WeylGroup.jl]: Thanks, I just realised myself during a seminar that I forgot about this :D

**Review by fingolfin — APPROVED** [review]
